### ISSUE TITLE
potential varLib bug in instantiation

### ISSUE THREAD

**Issue by laerm0** [googlefonts/gftools#143]
(@anthrotype [suggested](https://github.com/laerm0/opensans/issues/17) I file a bug)

When instantiating from a VTT-autohinted font, the composite autohints become broken.

Here's the variable font with its good composite low-level hint code:

```
/* VTTTalk Unicode 0xc9 */
/* ACT generated Thu Jun 20 15:25:27 2019 */

USEMYMETRICS[]
OFFSET[R], 38, 0, 0
OFFSET[R], 971, 435, 375
CALL[], 21, 22, 88
SVTCA[Y]
CALL[], 12, 0, 0, 0, 1616, 86
SHC[2], 1
CALL[], 9
CALL[], 12, 13, 11, 30, 10, 87
SVTCA[X]
SHC[2], 1
```

![screenshot_2019-07-02_10-36-04](https://user-images.githubusercontent.com/10780161/60521415-3d1f8400-9cb5-11e9-8411-89e7e634e3af.png)

Here's an instance at an axis maximum:

![screenshot_2019-07-02_10-36-36](https://user-images.githubusercontent.com/10780161/60521476-57596200-9cb5-11e9-80d2-a07aad792c92.png)

You can see that the extra-bold, the second glyph, is incorrect. It uses the same low-level hint code as the default instance.

I'll explain what the pertinent parts of the hint code are doing so you can see where the adjustments would have to be made.

`USEMYMETRICS[]`  
means this is a composite and I am specifying metrics  

`OFFSET[R], z, x, y`  
z = GID, x = X offset, y = Y offset  

`CALL[], 12, 13, 11, 30, 10, 87`  
Function call 87 centers an accent over a base. It is centering points 12 and 13 between points 11 and 10 with an offset of 30.

# whoa

I'm keeping this here so I don't lose my work and in case I do need it, but, as noted at https://github.com/laerm0/opensans/issues/17#issuecomment-507716129, I'm pretty sure the error is on Microsoft's end. I'm contacting them and will return to this if I need to. I am going to close as soon as I file this and re-open down the road if necessary.

**Comment by laerm0** [comment]
Nope, it had nothing to do with the hints. Remove the composite autohints and you still have misaligned accents, but only in the italic.

![screenshot_2019-07-04_15-15-54](https://user-images.githubusercontent.com/10780161/61559598-0b8a1500-aa38-11e9-9ce8-9754d2d968be.png)

Take a look at the [italic instances here](https://github.com/laerm0/opensans/tree/master/source/TTF%20VTT%20source) near the axis extrema and you'll see misaligned accents (i.e. the light and extrabold weights).

Seems to only be an issue in the italic, if that helps. Thanks.

**Comment by anthrotype** [comment]
What combo of browser + operating system is that screenshot from?
I can't seem to reproduce the issue using the italic instances you linked above (i tried light and extrabold weighs), I'm using latest Chrome on Linux.

Are you sure it is the precomposed accented glyphs and not the character combinations using combining marks and GPOS mark feature?
Maybe the rendering environment you are trying to render a text contiaining combining marks doesn't support GPOS mark feature?

**Comment by laerm0** [comment]
The screenshot of Vietnamese glyphs is Win 10 Chrome. 

If you open the fonts in any font editor (the examples at the top are VTT), you can see the accents are misaligned. It may also be related the the metrics of ligatures breaking post-instantiation.

![screenshot_2019-07-23_10-29-19](https://user-images.githubusercontent.com/10780161/61720490-e0077300-ad34-11e9-8baf-f8417391da6b.png)

**Comment by anthrotype** [comment]
please try be a bit more specific: which font file, which characters or glyphs

also, what is the expected output and how it is different from the actual output that you get.

**Comment by laerm0** [comment]
Here's a similar example from OpenSans-Italic-wg300wd75.ttf in OTMaster.

![screenshot_2019-07-23_10-33-15](https://user-images.githubusercontent.com/10780161/61720750-5f954200-ad35-11e9-89db-3c57ee43940b.png)

**Comment by laerm0** [comment]
Okay, well, start with the ligatures in the Italic `wg300wd75` and `wg800wd75` weights. The metrics are based on the first glyph in the ligature. Then take a look at the accented cap A glyphs. Those accents should be centered above the glyph, but many are misaligned (`uni1EB0` `uni1EB2` `uni1EA8` very much, `Aacute` `Abreve` `Amacron` more subtly off).

Thanks.

**Comment by laerm0** [comment]
The source TTFs are in that directory of the repo: `opensans/source/TTF VTT source/OpenSans-Italic.ttf`

You can also rebuild from source if you wish. The build script is in `opensans/source/`.

**Comment by anthrotype** [comment]
micah, can you try instantiating the _unhinted_ varible fonts and compare the result with instances generated from the variable font that went through VTT?
Do you see the same issues with component offsets? If not, I have a hunch that VTT is messing things up somehow

**Comment by anthrotype** [comment]
I get huge differences between the variable fonts generated using the source/build.sh script and those in "TTF VTT sources" (I am looking at the master branch). How were the latter varfonts generated? Were these not generated with the same build commands? What did they go through after that?

**Comment by laerm0** [comment]
I've just instantiated from freshly built fonts, unhinted, and you're right: these glyphs are fine, including the ligatures. Hmm. I'll have to ponder this and see what the hinting is breaking.

**Comment by laerm0** [comment]
> I get huge differences between the variable fonts generated using the source/build.sh script and those in "TTF VTT sources" (I am looking at the master branch). 

What kinds of diffs are you seeing?

> How were the latter varfonts generated? Were these not generated with the same build commands? What did they go through after that?

These files were generated using the same build process but I had to do a lot of table tweaking to the TTFs so that they would accept the old hints from the original hinted files. The *very ugly* process is documented at https://github.com/laerm0/opensans/blob/master/source/instance_creation_process.md .

**Comment by anthrotype** [comment]
difference are too many. Try yourself to rebuild the unhinted variable font, dump to ttx, then dump your VTT-hinted variable font to ttx (I took the one I found in that "TTF VTT sources" folder), and diff the two.
It's not just the added hinting data. Even the glyf outlines show differences.
For example, take the first contour on letter "A". It's the same points but the contour direction is flipped!
Did you maybe use --keep-direction option when you produced the varfont before starting the manual hinting in VTT? The build.sh script is not using that.

```xml
     <TTGlyph name="A" xMin="0" yMin="0" xMax="1229" yMax="1468">
       <contour>
         <pt x="1229" y="0" on="1"/>
-        <pt x="653" y="1468" on="1"/>
-        <pt x="588" y="1468" on="1"/>
-        <pt x="0" y="0" on="1"/>
-        <pt x="113" y="0" on="1"/>
-        <pt x="317" y="516" on="1"/>
-        <pt x="911" y="516" on="1"/>
         <pt x="1114" y="0" on="1"/>
+        <pt x="911" y="516" on="1"/>
+        <pt x="317" y="516" on="1"/>
+        <pt x="113" y="0" on="1"/>
+        <pt x="0" y="0" on="1"/>
+        <pt x="588" y="1468" on="1"/>
+        <pt x="653" y="1468" on="1"/>
       </contour>
```

**Comment by laerm0** [comment]
Well, when I originally hinted these files, I was unaware of fontmake and the build process was essentially just have Glyphs export the VFs. That's why the contour direction on the Italic is wrong. I had to do so much work on the Roman after the fact, though, that I eventually just rebuilt using the proper procedures and so it has a correct contour direction.

At this point, maybe I'm just going to rehint the italic from scratch. Yuck. I'll see what I can figure out.

**Comment by davelab6** [comment]
@laerm0  Was there a varlib bug here? :) If not, please close
